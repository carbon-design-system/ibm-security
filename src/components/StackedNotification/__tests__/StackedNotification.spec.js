/**
 * @file Stacked notification tests.
 * @copyright IBM Security 2019
 */

import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { StackedNotification } from '../../..';

describe('StackedNotification', () => {
  test('should have no Axe or DAP violations', async () => {
    const main = document.createElement('main');
    render(
      <StackedNotification
        title="test title"
        subtitle="test subtitle"
        caption="test caption"
        iconDescription="test close button icon"
      />,
      {
        // DAP requires a landmark '<main>' in the DOM:
        container: document.body.appendChild(main),
      }
    );
    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations('StackedNotification');
  });

  test('should add a custom class', () => {
    const { container } = render(
      <StackedNotification
        title="test title"
        className="custom-class"
        subtitle="test subtitle"
        iconDescription="test icon desc"
      />
    );
    expect(container.querySelector('.custom-class')).toBeVisible();
  });

  test('should invoke close mock when close button is clicked', () => {
    const onCloseMock = jest.fn();
    const { getByLabelText } = render(
      <StackedNotification
        title="test title"
        subtitle="test subtitle"
        iconDescription="test close button icon"
        onCloseButtonClick={onCloseMock}
      />
    );
    userEvent.click(getByLabelText(/test close button icon/i));
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  test('should pass through extra props via spread attribute', () => {
    const { queryByTestId } = render(
      <StackedNotification
        title="test title"
        subtitle="test subtitle"
        iconDescription="test close button icon"
        data-testid="test-id"
      />
    );
    expect(queryByTestId('test-id')).toBeVisible();
  });

  test('should apply a `title`', () => {
    const { queryByText } = render(
      <StackedNotification
        title="test title"
        subtitle="test subtitle"
        iconDescription="test close button icon"
      />
    );
    expect(queryByText(/test title/i)).toBeVisible();
  });

  test('should apply a `subtitle`', () => {
    const { queryByText } = render(
      <StackedNotification
        title="test title"
        subtitle="test subtitle"
        iconDescription="test close button icon"
      />
    );
    expect(queryByText(/test subtitle/i)).toBeVisible();
  });

  test('should apply a `caption`', () => {
    const { queryByText } = render(
      <StackedNotification
        title="test title"
        subtitle="test subtitle"
        iconDescription="test close button icon"
        caption="test caption"
      />
    );
    expect(queryByText(/test caption/i)).toBeVisible();
  });

  test('should apply a `children`', () => {
    const { queryByText } = render(
      <StackedNotification
        title="test title"
        subtitle="test subtitle"
        iconDescription="test close button icon"
      >
        test children
      </StackedNotification>
    );
    expect(queryByText(/test children/i)).toBeVisible();
  });

  test('should apply a custom `role`', () => {
    const { queryByRole } = render(
      <StackedNotification
        title="test title"
        subtitle="test subtitle"
        iconDescription="test close button icon"
        role="dialog"
      />
    );
    // Note that "alert" is the default role:
    expect(queryByRole('dialog')).toBeVisible();
  });

  test('should apply an `iconDescription` to the close button', () => {
    const { queryByLabelText } = render(
      <StackedNotification
        title="test title"
        subtitle="test subtitle"
        iconDescription="test close button icon"
      />
    );
    expect(queryByLabelText(/test close button icon/i)).toBeVisible();
  });

  test('should hide the close button when `hideCloseButton` is `true`', () => {
    const { queryByLabelText } = render(
      <StackedNotification
        title="test title"
        subtitle="test subtitle"
        iconDescription="test close button icon"
        hideCloseButton
      />
    );
    expect(queryByLabelText(/test close button icon/i)).not.toBeInTheDocument();
  });
});
