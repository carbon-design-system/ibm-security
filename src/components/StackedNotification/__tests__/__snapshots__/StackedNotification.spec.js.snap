// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackedNotification render renders correctly 1`] = `
<ToastNotification
  caption=""
  className="security--stacked-notification"
  hideCloseButton={false}
  iconDescription="Close notification"
  kind="info"
  notificationType="toast"
  onClick={[Function]}
  onCloseButtonClick={[Function]}
  role="alert"
  subtitle="Roius abnta mod tempor adsfiso incidfid idunt ut labnt empor uilt labnore..."
  timeout={0}
  title="You have 4 unread notifications"
/>
`;

exports[`StackedNotification render renders the HTML of the node's subtree 1`] = `
<div
  class="bx--toast-notification bx--toast-notification--info security--stacked-notification"
  kind="info"
  role="alert"
  timeout="0"
>
  <div
    class="bx--toast-notification__details"
  >
    <h3
      class="bx--toast-notification__title"
    >
      You have 4 unread notifications
    </h3>
    <div
      class="bx--toast-notification__subtitle"
    >
      Roius abnta mod tempor adsfiso incidfid idunt ut labnt empor uilt labnore...
    </div>
    <div
      class="bx--toast-notification__caption"
    />
  </div>
  <button
    aria-label="Close notification"
    class="bx--toast-notification__close-button"
    title="Close notification"
    type="button"
  >
    <svg
      aria-label="close notification"
      class="bx--toast-notification__close-icon"
      focusable="false"
      height="20"
      preserveAspectRatio="xMidYMid meet"
      role="img"
      style="will-change:transform"
      viewBox="0 0 32 32"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
      />
    </svg>
  </button>
</div>
`;
